core/offchain/impl/runner.hpp

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_OFFCHAIN_IMPL_RUNNER_HPP
+#define KAGOME_OFFCHAIN_IMPL_RUNNER_HPP
+
+#include <deque>
+#include <gsl/gsl_util>
+#include <mutex>
+
+#include "utils/thread_pool.hpp"
+
+namespace kagome::offchain {
+  /**
+   * Enqueue at most `max_tasks_` to run on number of `threads_`.
+   * Old tasks do not run and are removed when queue is full.
+   */
+  class Runner : public std::enable_shared_from_this<Runner> {
+   public:
+    using Task = std::function<void()>;
+
+    Runner(size_t threads, size_t max_tasks)
+        : threads_{threads},
+          free_threads_{threads},
+          max_tasks_{max_tasks},
+          thread_pool_{threads_} {}
+
+    void run(Task &&task) {
+      std::unique_lock lock{mutex_};
+      if (tasks_.size() >= max_tasks_) {
+        tasks_.pop_front();
+      }
+      if (free_threads_ <= 0) {
```

Comment:
`free_threads_` can not be negative
```suggestion
      if (free_threads_ == 0) {
```